src/librustc_mir_build/hair/pattern/const_to_pat.rs

```diff
@@ -257,7 +238,87 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
                 slice: None,
                 suffix: Vec::new(),
             },
-            _ => PatKind::Constant { value: cv },
+            ty::Ref(_, pointee_ty, ..) if pointee_ty.is_sized(tcx.at(self.span), param_env) => {
+                let old = self.backcompat_hack.replace(true);
+                let val = PatKind::Deref {
+                    subpattern: self.recur(tcx.deref_const(self.param_env.and(cv))),
+                };
+                self.backcompat_hack.set(old);
+                val
+            }
+            // `str`, `[T]`, extern types and dyn Trait.
+            ty::Ref(_, pointee_ty, ..) => match pointee_ty.kind {
+                // `&str` and `&[u8]` are represented as `ConstValue::Slice`, let's keep using this
+                // optimization for now.
+                ty::Str => PatKind::Constant { value: cv },
```

Comment:
I think `&str` is now the only one that I don't know how to handle (i know, `&[u8]` also isn't handled yet). Maybe we could create a new `PatKind` just for it? What do you think?

Reply:
alternatively we could treat `&[char]` and `&str` structurally equal. Not sure if I'm serious about that

Reply:
cc @Nadrieril

Reply:
Do you mean you don't know how to get the underlying `&[u8]` ? The PatKind would be opaque then, like all constants currently are ? Yeah, sounds legit. A generic `PatKind::OpaqueConstant` might actually be needed anyways if I understand @petrochenkov's remark about `#[structural_match]`